From 32203e71dfc00dc4f350cea6645b912038da563d Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Sun, 17 Jul 2016 15:27:12 +0300
Subject: [PATCH 0380/1345] pp3: nss0 mtu change support

commit  59fa90fec61e30cc7874c9d2c3663a5f6d418b95 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Permit MTU change for nss-default-gateway nss0
with RX-buffer-size setting according to the same MTU
(like for emac), but without FW emac.mtu message sending.

Other nssN are restricted.

Change-Id: Icc6f269b22d9cfaf45e175e53307d0577b63c798
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/31176
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Dmitri Epshtein <dima@marvell.com>
Tested-by: Dmitri Epshtein <dima@marvell.com>
Reviewed-by: Dovrat Zifroni <dovrat@marvell.com>
Verified-Armada8K: Dovrat Zifroni <dovrat@marvell.com>
Tested-by: Dovrat Zifroni <dovrat@marvell.com>
Reviewed-by: Yelena Krivosheev <yelena@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../net/ethernet/marvell/pp3/net_dev/mv_netdev.c   |   13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c b/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
index 5b7d8d0..608cd08 100644
--- a/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
+++ b/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
@@ -2932,8 +2932,10 @@ static int mv_pp3_change_mtu_internals(struct net_device *dev, int mtu)
 			pr_warn("%s: FW long pool update failed\n", __func__);
 	}
 
-	emac_vp = dev_priv->vport;
+	if (dev_priv->vport->type != MV_PP3_NSS_PORT_ETH)
+		return 0;
 
+	emac_vp = dev_priv->vport;
 	emac_vp->port.emac.mtu = mtu;
 
 	if (pp3_fw_vport_mtu_set(emac_vp->vport, mtu) < 0)
@@ -2961,9 +2963,12 @@ static int mv_pp3_change_mtu(struct net_device *dev, int mtu)
 		return 0;
 	}
 
-	/* Supported only for EMAC virtual ports */
-	if (!dev_priv->vport || (dev_priv->vport->type != MV_PP3_NSS_PORT_ETH)) {
-		pr_err("%s: not supported for %s\n", __func__, dev->name);
+	if (!dev_priv->vport)
+		goto error;
+
+	if ((dev_priv->vport->type != MV_PP3_NSS_PORT_ETH) &&
+	    (dev_priv->id != MV_NSS_EXT_PORT_MAX)) {
+		pr_err("%s: MTU change is irrelevant for %s\n", dev->name, dev->name);
 		return -1;
 	}
 
-- 
1.7.9.5

